### PR TITLE
feat: initialize file logging via env

### LINKED ISSUE #14705: Initialize file logging to more early
### What would you like Renovate to be able to do?

Currently we miss a lot of log messages on file logging here
https://github.com/renovatebot/renovate/blob/dca3418bbdbf1024ac6807e591562b4042bc846d/lib/workers/global/config/parse/index.ts#L58-L70

I like to remove the config oiptions and initialize it via env only, like we do for `LOG_LEVEL`.

### If you have any ideas on how this should be implemented, please tell us here.

Sure, do it like we do for `LOG_LEVEL`. Deprecate old option and remove in a future major change.

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
Initialize file logging at the very start if environment variables `LOG_FILE` and `LOG_FILE_LEVEL` are set. And, if these variables are set make sure to skip the file logging logic present in [config/parse/index
](https://github.com/renovatebot/renovate/blob/dca3418bbdbf1024ac6807e591562b4042bc846d/lib/workers/global/config/parse/index.ts#L58-L70)

## Context
Closes: https://github.com/renovatebot/renovate/issues/14705

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or (pending)
- [ ] No documentation update is required 

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/logger/index.ts [hunk 1]
```diff
@@ -123,6 +125,20 @@ loggerLevels.forEach((loggerLevel) => {
   logger.once[loggerLevel] = logOnceFn as never;
 });
 
+if (is.string(process.env.LOG_FILE)) {
+  validateLogLevel(process.env.LOG_FILE_LEVEL);
+
+  // ensure log file directory exists
+  const directoryName = upath.parse(process.env.LOG_FILE).dir;
```
Comment (viceice) [inline lib/logger/index.ts:132]:
```suggestion
  const directoryName = upath.dirname(process.env.LOG_FILE);
```
isn't that the same?

File: lib/logger/index.ts [hunk 2]
```diff
@@ -123,6 +125,20 @@ loggerLevels.forEach((loggerLevel) => {
   logger.once[loggerLevel] = logOnceFn as never;
 });
 
+if (is.string(process.env.LOG_FILE)) {
+  validateLogLevel(process.env.LOG_FILE_LEVEL);
```
Comment (viceice) [inline lib/logger/index.ts:129]: can we return the validated level here and use it below, so we don't need the unsafe cast?

**Review by viceice — COMMENTED** [review]
we should log a deprecation warning for the old config option and remove them from docs.

update logging docs with new ENV config.

File: docs/usage/config-overview.md [hunk 3]
```diff
@@ -110,6 +110,8 @@ Finally, there are some special environment variables that are loaded _before_ c
 - `LOG_CONTEXT`: a unique identifier used in each log message to track context
 - `LOG_FORMAT`: defaults to a "pretty" human-readable output, but can be changed to "json"
 - `LOG_LEVEL`: most commonly used to change from the default `info` to `debug` logging
+- `LOG_FILE`: used to specify the log file path
+- `LOG_FILE_LEVEL`: log file logging level, defaults to `info`
```
Comment (rarkins) [inline docs/usage/config-overview.md:114]: Keep alphabetical order

**Review by rarkins — CHANGES_REQUESTED** [review]
It should not be _moved_, and file-based config.js should not be removed.

Instead,
- If LOG_FILE is defined then it should be initialized early
- Keep existing initialization for config.logLevel but skip it if it was already initialized due to presence of process.env.LOG_LEVEL

File: lib/logger/utils.ts [hunk 4]
```diff
@@ -277,27 +277,30 @@ export function withSanitizer(streamConfig: bunyan.Stream): bunyan.Stream {
...
+
   if (
     is.undefined(logLevelToCheck) ||
     (is.string(logLevelToCheck) &&
-      allowedValues.includes(logLevelToCheck as bunyan.LogLevel))
+      allowedValues.includes(logLevelToCheck as bunyan.LogLevelString))
   ) {
     // log level is in the allowed values or its undefined
-    return;
+    return (logLevelToCheck as bunyan.LogLevelString) ?? 'info';
```
Comment (viceice) [inline lib/logger/utils.ts:303]: always return info as fallback

File: lib/logger/index.ts [hunk 5]
```diff
@@ -20,12 +22,10 @@ if (is.string(process.env.LOG_LEVEL)) {
...
 }
 
-validateLogLevel(process.env.LOG_LEVEL);
+const logLevel = validateLogLevel(process.env.LOG_LEVEL);
 const stdout: bunyan.Stream = {
   name: 'stdout',
-  level:
-    (process.env.LOG_LEVEL as bunyan.LogLevel) ||
-    /* istanbul ignore next: not testable */ 'info',
+  level: logLevel || /* istanbul ignore next: not testable */ 'info',
```
Comment (viceice) [inline lib/logger/index.ts:28]:
```suggestion
  level: logLevel,
```

File: docs/usage/self-hosted-configuration.md [hunk 6]
```diff
@@ -729,10 +729,6 @@ When you set `inheritConfigStrict=true` then Renovate will abort the run and rai
 `logContext` is included with each log entry only if `logFormat="json"` - it is not included in the pretty log output.
 If left as default (null), a random short ID will be selected.
 
-## logFile
-
-## logFileLevel
-
```
Comment (viceice) [inline docs/usage/self-hosted-configuration.md:735]: I think we need to keep it to pass docs validation test 🤔

**Comment by RahulGautamSingh** [comment]
> It should not be _moved_, and file-based config.js should not be removed.

 What about warning that config options `logFile` and `logFileLevel` are now deprecated?

File: docs/usage/config-overview.md [hunk 7]
```diff
@@ -108,6 +108,8 @@ Read the [Self-hosted experimental environment variables](./self-hosted-experime
 Finally, there are some special environment variables that are loaded _before_ configuration parsing because they are used during logging initialization:
 
 - `LOG_CONTEXT`: a unique identifier used in each log message to track context
+- `LOG_FILE`: used to specify the log file path
+- `LOG_FILE_LEVEL`: log file logging level, defaults to `info`
```
Comment (rarkins) [inline docs/usage/config-overview.md:112]: I prefer it falls back to debug like today

File: lib/config/options/index.ts [hunk 8]
```diff
@@ -484,14 +484,14 @@ const options: RenovateOptions[] = [
   // Log options
   {
     name: 'logFile',
-    description: 'Log file path.',
+    description: 'Log file path. (deprecated)',
```
Comment (viceice) [inline lib/config/options/index.ts:487]: We should add a `deprecated` boolean field, which automatically render a warning to the docs, like we do for `experimental`

https://github.com/renovatebot/renovate/blob/a6f90f26698d21db64f2732cb114566a4d3d1d92/tools/docs/config.ts#L157

File: lib/config/validation.spec.ts [hunk 9]
```diff
@@ -1223,6 +1223,19 @@ describe('config/validation', () => {
...
+      const config = {
+        logFile: 'something',
+        logFileLevel: 'debug',
+      };
+      const { warnings } = await configValidation.validateConfig(
+        'global',
+        config,
+      );
+      expect(warnings).toHaveLength(2);
+      expect(warnings).toMatchSnapshot();
```
Comment (viceice) [inline lib/config/validation.spec.ts:1236]: don't use snapshot here.

File: lib/logger/index.ts [hunk 10]
```diff
@@ -20,12 +22,10 @@ if (is.string(process.env.LOG_LEVEL)) {
...
 }
 
-validateLogLevel(process.env.LOG_LEVEL);
+const logLevel = validateLogLevel(process.env.LOG_LEVEL);
 const stdout: bunyan.Stream = {
   name: 'stdout',
-  level:
-    (process.env.LOG_LEVEL as bunyan.LogLevel) ||
-    /* istanbul ignore next: not testable */ 'info',
+  level: logLevel,
```
Comment (viceice) [inline lib/logger/index.ts:28]:
```suggestion
  level: validateLogLevel(process.env.LOG_LEVEL),
```

File: lib/logger/index.ts [hunk 11]
```diff
@@ -20,12 +22,10 @@ if (is.string(process.env.LOG_LEVEL)) {
   process.env.LOG_LEVEL = process.env.LOG_LEVEL.toLowerCase().trim();
 }
 
-validateLogLevel(process.env.LOG_LEVEL);
+const logLevel = validateLogLevel(process.env.LOG_LEVEL);
```
Comment (viceice) [inline lib/logger/index.ts:25]:
```suggestion
```

File: docs/usage/config-overview.md (hunk 7 above)
Comment (viceice) [reply docs/usage/config-overview.md:112]: should renovate `LOG_LEVLE` also default to `debug` then?

@RahulGautamSingh If we need different levels, then add an optional default value. `coerceString` can help 🤗

File: lib/logger/index.ts [hunk 12]
```diff
@@ -123,6 +123,21 @@ loggerLevels.forEach((loggerLevel) => {
...
+  const logFileLevel = validateLogLevel(process.env.LOG_FILE_LEVEL);
+
+  // ensure log file directory exists
+  const directoryName = upath.dirname(process.env.LOG_FILE);
+  fs.ensureDirSync(directoryName);
+
+  addStream({
+    name: 'logfile',
+    path: process.env.LOG_FILE,
+    level: logFileLevel,
```
Comment (viceice) [inline lib/logger/index.ts:137]:
```suggestion
    level: validateLogLevel(process.env.LOG_FILE_LEVEL),
```

File: lib/logger/index.ts [hunk 13]
```diff
@@ -123,6 +123,21 @@ loggerLevels.forEach((loggerLevel) => {
   logger.once[loggerLevel] = logOnceFn as never;
 });
 
+// istanbul ignore if: not easily testable
+if (is.string(process.env.LOG_FILE)) {
+  const logFileLevel = validateLogLevel(process.env.LOG_FILE_LEVEL);
+
```
Comment (viceice) [inline lib/logger/index.ts:129]:
```suggestion
```

File: docs/usage/config-overview.md (hunk 7 above)
Comment (RahulGautamSingh) [reply docs/usage/config-overview.md:112]: >I prefer it falls back to debug like today
This is confusing, because the [docs](https://docs.renovatebot.com/config-overview/#logging-variables) say, we fall back to `info`.

File: lib/workers/global/config/parse/index.spec.ts [hunk 14]
```diff
@@ -1,13 +1,13 @@
...
 import { mocked } from '../../../../../test/util';
+import { logger } from '../../../../logger';
 import { readSystemFile } from '../../../../util/fs';
 import getArgv from './__fixtures__/argv';
 import * as _hostRulesFromEnv from './host-rules-from-env';
 
 jest.mock('../../../../modules/datasource/npm');
 jest.mock('../../../../util/fs');
 jest.mock('./host-rules-from-env');
-jest.mock('../../config.js', () => ({}), { virtual: true });
```
[the remark this answers is not in the document]
Comment (RahulGautamSingh) [reply lib/workers/global/config/parse/index.spec.ts:10]: It didn't. The mock is for the file to exist in the `lib/workers/global/` folder, whereas locally we store the file in the root dir.

File: lib/config/options/index.ts (hunk 8 above)
Comment (RahulGautamSingh) [reply lib/config/options/index.ts:487]: https://github.com/renovatebot/renovate/pull/28216

**Comment by RahulGautamSingh** [comment]
Blocked by: https://github.com/renovatebot/renovate/pull/28216
Need the above PR merged so that the `logFile` and `logFileLevel` can be deprecated in a good manner.

File: docs/usage/config-overview.md (hunk 7 above)
Comment (rarkins) [reply docs/usage/config-overview.md:112]: stdout logging should default to info, file logging default to debug

**Comment by rarkins** [comment]
https://github.com/renovatebot/renovate/issues/14705#issuecomment-2047312905

**Comment by viceice** [comment]
> [#14705 (comment)](https://github.com/renovatebot/renovate/issues/14705#issuecomment-2047312905)

We should simply warn here and drop the old config option in v38

File: lib/logger/utils.ts [hunk 15]
```diff
@@ -277,27 +277,30 @@ export function withSanitizer(streamConfig: bunyan.Stream): bunyan.Stream {
...
- * A function that terminates exeution if the log level that was entered is
+ * A function that terminates execution if the log level that was entered is
  *  not a valid value for the Bunyan logger.
  * @param logLevelToCheck
  * @returns returns undefined when the logLevelToCheck is valid. Else it stops execution.
  */
-export function validateLogLevel(logLevelToCheck: string | undefined): void {
-  const allowedValues: bunyan.LogLevel[] = [
+export function validateLogLevel(
+  logLevelToCheck: string | undefined,
```
Comment (viceice) [inline lib/logger/utils.ts:286]:
```suggestion
  logLevelToCheck: string | undefined,
  defaultLevel?: bunyan.LogLevelString
```

File: lib/logger/utils.ts (hunk 4 above)
Comment (viceice) [inline lib/logger/utils.ts:303]:
```suggestion
    return (logLevelToCheck as bunyan.LogLevelString) ?? defaultLevel ?? 'info';
```

**Comment by rarkins** [comment]
Retargeted to v38 branch. v38 needs rebasing

File: lib/config/validation.ts [hunk 16]
```diff
@@ -95,6 +95,8 @@ function getDeprecationMessage(option: string): string | undefined {
     branchName: `Direct editing of branchName is now deprecated. Please edit branchPrefix, additionalBranchPrefix, or branchTopic instead`,
     commitMessage: `Direct editing of commitMessage is now deprecated. Please edit commitMessage's subcomponents instead.`,
     prTitle: `Direct editing of prTitle is now deprecated. Please edit commitMessage subcomponents instead as they will be passed through to prTitle.`,
+    logFile: `Using logFile to specify log file name is deprecated now. Please use the enviroment variable LOG_FILE instead`,
+    logFileLevel: `Using logFileLevel to specify log file name is deprecated now. Please use the enviroment variable LOG_FILE_LEVEL instead`,
```
Comment (RahulGautamSingh) [inline lib/config/validation.ts:99]:
```suggestion
    logFileLevel: `Using logFileLevel to specify log level for file logging is deprecated now. Please use the enviroment variable LOG_FILE_LEVEL instead`,
```

**Review by viceice — DISMISSED** [review]

File: docs/usage/config-overview.md [hunk 17]
```diff
@@ -108,6 +108,8 @@ Read the [Self-hosted experimental environment variables](./self-hosted-experime
 Finally, there are some special environment variables that are loaded _before_ configuration parsing because they are used during logging initialization:
 
 - `LOG_CONTEXT`: a unique identifier used in each log message to track context
+- `LOG_FILE`: used to specify the log file path
```
Comment (rarkins) [inline docs/usage/config-overview.md:111]:
```suggestion
- `LOG_FILE`: used to enable file logging and specify the log file path
```

File: lib/config/validation.ts [hunk 18]
```diff
@@ -95,6 +95,8 @@ function getDeprecationMessage(option: string): string | undefined {
     branchName: `Direct editing of branchName is now deprecated. Please edit branchPrefix, additionalBranchPrefix, or branchTopic instead`,
     commitMessage: `Direct editing of commitMessage is now deprecated. Please edit commitMessage's subcomponents instead.`,
     prTitle: `Direct editing of prTitle is now deprecated. Please edit commitMessage subcomponents instead as they will be passed through to prTitle.`,
+    logFile: `Using logFile to specify log file name is deprecated now. Please use the enviroment variable LOG_FILE instead`,
+    logFileLevel: `Using logFileLevel to specify log level for file logging is deprecated now. Please use the enviroment variable LOG_FILE_LEVEL instead`,
```
Comment (rarkins) [inline lib/config/validation.ts:99]: Do we print both these messages and the deprecation ones? Are they printed for global config too?

File: lib/config/validation.ts [hunk 19]
```diff
@@ -851,6 +853,12 @@ async function validateGlobalConfig(
   currentPath: string | undefined,
   config: RenovateConfig,
 ): Promise<void> {
+  if (getDeprecationMessage(key)) {
+    warnings.push({
+      topic: 'Deprecation Warning',
+      message: getDeprecationMessage(key)!,
+    });
+  }
```
Comment (rarkins) [inline lib/config/validation.ts:861]: Why was this change needed in this PR?

File: lib/logger/index.ts [hunk 20]
```diff
@@ -20,12 +22,9 @@ if (is.string(process.env.LOG_LEVEL)) {
   process.env.LOG_LEVEL = process.env.LOG_LEVEL.toLowerCase().trim();
 }
 
-validateLogLevel(process.env.LOG_LEVEL);
 const stdout: bunyan.Stream = {
   name: 'stdout',
-  level:
-    (process.env.LOG_LEVEL as bunyan.LogLevel) ||
-    /* istanbul ignore next: not testable */ 'info',
+  level: validateLogLevel(process.env.LOG_LEVEL),
```
Comment (rarkins) [inline lib/logger/index.ts:27]:
```suggestion
  level: validateLogLevel(process.env.LOG_LEVEL, 'info'),
```

File: lib/logger/utils.ts [hunk 21]
```diff
@@ -277,27 +277,31 @@ export function withSanitizer(streamConfig: bunyan.Stream): bunyan.Stream {
...
+  defaultLevel?: bunyan.LogLevelString,
```
Comment (rarkins) [inline lib/logger/utils.ts:287]:
```suggestion
  defaultLevel: bunyan.LogLevelString,
```

File: lib/logger/utils.ts [hunk 22]
```diff
@@ -277,27 +277,31 @@ export function withSanitizer(streamConfig: bunyan.Stream): bunyan.Stream {
...
+
   if (
     is.undefined(logLevelToCheck) ||
     (is.string(logLevelToCheck) &&
-      allowedValues.includes(logLevelToCheck as bunyan.LogLevel))
+      allowedValues.includes(logLevelToCheck as bunyan.LogLevelString))
   ) {
     // log level is in the allowed values or its undefined
-    return;
+    return (logLevelToCheck as bunyan.LogLevelString) ?? defaultLevel ?? 'info';
```
Comment (rarkins) [inline lib/logger/utils.ts:304]:
```suggestion
    return (logLevelToCheck as bunyan.LogLevelString) ?? defaultLevel;
```

**Review by rarkins — CHANGES_REQUESTED** [review]
Isn't this now non-breaking, so doesn't need to target v38?

File: lib/config/validation.ts (hunk 18 above)
Comment (RahulGautamSingh) [reply lib/config/validation.ts:99]: These are the depreciation msgs which will be shown in the logs. If you mean to ask whether there will be 2 log messages, the answer is depends. 
If the logFile / logFileContext are of wrong types. Then there willl be two warning logs, one for depriciation and another for type mismatch else onlt the depreciation logs.

File: lib/workers/global/config/parse/index.spec.ts [hunk 23]
```diff
@@ -176,7 +176,7 @@ describe('workers/global/config/parse/index', () => {
 
     it('massage onboardingNoDeps when autodiscover is false', async () => {
       jest.mock(
```
Comment (viceice) [inline lib/workers/global/config/parse/index.spec.ts:178]:
```suggestion
       jest.doMock(
```

File: lib/workers/global/config/parse/index.spec.ts [hunk 24]
```diff
@@ -186,5 +186,38 @@ describe('workers/global/config/parse/index', () => {
       const parsedConfig = await configParser.parseConfigs(env, defaultArgv);
       expect(parsedConfig).toContainEntries([['onboardingNoDeps', 'enabled']]);
     });
+
+    it('initalizes file logging when logFile is set and env vars LOG_FILE is undefined', async () => {
+      jest.mock(
```
Comment (viceice) [inline lib/workers/global/config/parse/index.spec.ts:191]:
```suggestion
      jest.doMock(
```

File: lib/workers/global/config/parse/index.spec.ts [hunk 25]
```diff
@@ -186,5 +186,38 @@ describe('workers/global/config/parse/index', () => {
...
+      const env: NodeJS.ProcessEnv = {};
+      const parsedConfig = await configParser.parseConfigs(env, defaultArgv);
+      expect(parsedConfig).not.toContain([['logFile', 'someFile']]);
+      expect(logger.debug).toHaveBeenCalledWith(
+        'Enabling debug logging to somepath',
+      );
+    });
+
+    it('skips initializing file logging when logFile is set but env vars LOG_FILE is defined', async () => {
+      jest.mock(
```
Comment (viceice) [inline lib/workers/global/config/parse/index.spec.ts:207]:
```suggestion
      jest.doMock(
```

File: lib/workers/global/config/parse/index.spec.ts [hunk 26]
```diff
@@ -174,17 +173,54 @@ describe('workers/global/config/parse/index', () => {
...
-        },
-      );
-      const env: NodeJS.ProcessEnv = {};
-      const parsedConfig = await configParser.parseConfigs(env, defaultArgv);
-      expect(parsedConfig).toContainEntries([['onboardingNoDeps', 'enabled']]);
+    describe('haha', () => {
+      beforeEach(() => {
+        jest.resetModules();
+      });
+
```
Comment (viceice) [inline lib/workers/global/config/parse/index.spec.ts:180]:
```suggestion
```
I don't think you need this

**Review by rarkins — DISMISSED** [review]

**Review by viceice — APPROVED** [review]
